Library/Homebrew/upgrade.rb

```diff
@@ -209,7 +209,7 @@ def install_formula(formula_installer, upgrade:)
     ensure
       # restore previous installation state if build failed
       begin
-        linked_kegs.each(&:link) if linked_kegs.present? && !f.latest_version_installed?
```

Comment:
The `f` here always results in a `NameError` here, afaitct. I believe this was a bug that was introduced in the beginning: https://github.com/Homebrew/brew/pull/12024/files#diff-338a6a0ff7eb3b254fa3b6ca7d9225e7e283521c05acbdcda7f4eb8dc996bc05R206

(You should be able to replicate this by removing the `rescue` and running `test/cmd/install_spec`.)

However: 1) it wasn't covered by tests (it seems `linked_kegs` is always `present?`), 2) it's wrapped in a bare `rescue` (which I would discourage, but I don't have sufficient context here to know how to better target it), and 3) the file was untyped.

`formula` is my best guess as to the original intent, and I've rearranged things a bit so that it's covered in tests.

cc @FnControlOption @MikeMcQuaid